lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,468 @@
+/// `Source` node to nodes that are successor of it on the spanning tree.
+template<class GraphT, class GT>
+CycleEquivalenceAnalysis<GraphT, GT>::EdgesVectorType
+CycleEquivalenceAnalysis<GraphT, GT>::getForwardEdges(BlockNode *Source) {
+  llvm::SmallVector<EdgeDescriptor> ForwardEdges;
+  for (auto &[Neighbor, Index] :
+       llvm::children_edges<llvm::Undirected<BlockNode *>>(Source)) {
+    EdgeDescriptor Edge = EdgeDescriptor(Source, Neighbor);
+    if (SpanningTreeEdges.contains(Index) and not isBackedge(Edge)) {
+      ForwardEdges.push_back(Edge);
+    }
+  }
+  return ForwardEdges;
+}
+
+/// Helper function that returns only the `backward` edges, according to the
+/// spanning tree. Backward edges are defined as edges connecting the current
+/// `Source` node to nodes that are not successors of it on the spanning tree,
+/// and therefore are backedges.
+template<class GraphT, class GT>
+CycleEquivalenceAnalysis<GraphT, GT>::EdgesVectorType
+CycleEquivalenceAnalysis<GraphT, GT>::getBackwardEdges(BlockNode *Source) {
+  llvm::SmallVector<EdgeDescriptor> BackwardEdges;
+  for (auto &[Neighbor, Index] :
+       llvm::children_edges<llvm::Undirected<BlockNode *>>(Source)) {
+    EdgeDescriptor Edge = EdgeDescriptor(Source, Neighbor);
+    if (not SpanningTreeEdges.contains(Index) and isBackedge(Edge)) {
+      BackwardEdges.push_back(Edge);
+    }
+  }
+  return BackwardEdges;
+}
+
+/// Helper function that returns only the `tree` edges, according to the
+/// spanning tree. Tree edges are defined as edges connecting the current
+/// `Source` node to nodes that are predecessor of it on the spanning tree.
+template<class GraphT, class GT>
+CycleEquivalenceAnalysis<GraphT, GT>::EdgesVectorType
+CycleEquivalenceAnalysis<GraphT, GT>::getTreeEdges(BlockNode *Source) {
+  llvm::SmallVector<EdgeDescriptor> TreeEdges;
+  for (auto &[Neighbor, Index] :
+       llvm::children_edges<llvm::Undirected<BlockNode *>>(Source)) {
+    EdgeDescriptor Edge = EdgeDescriptor(Neighbor, Source);
+    if (SpanningTreeEdges.contains(Index) and not isBackedge(Edge)) {
+      TreeEdges.push_back(Edge);
+    }
+  }
+  return TreeEdges;
+}
+
+// Helper function that creates the `GenericGraph` representing the undirected
+// version of the original CFG
+template<class GraphT, class GT = llvm::GraphTraits<GraphT>>
+static CycleEquivalenceAnalysis<GraphT>::BlockGraph
+initializeGenericGraph(GraphT F) {
+  using NodeT = GT::NodeRef;
+
+  using BlockNode = CycleEquivalenceAnalysis<GraphT>::BlockNode;
+  using BlockGraph = CycleEquivalenceAnalysis<GraphT>::BlockGraph;
+
+  BlockGraph Graph;
+
+  std::map<NodeT, BlockNode *> BBToNode;
+
+  // For every BasicBlock in F, we create a corresponding node in the
+  // `GenericGraph`, and we keep a correspondence between the `BasicBlock` and
+  // the node in the `GenericGraph`
+  auto NodesRange = llvm::make_range(GT::nodes_begin(F), GT::nodes_end(F));
+  for (NodeT BB : NodesRange) {
```

Comment:
Doesn't `llvm::nodes(F)` work?

Reply:
Yeah, I changed it to use `llvm::nodes` here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L319, but I am a bit conflicted about it. This method (and the class indeed), takes a second template parameter, being the `GraphTraits` telling how to iterate over the original graph. Using `llvm::nodes` prevents this opportunity, while the previous one explicitly used them (as it is done when iterating over the `children`, cfr. here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L338.

Reply:
Graph traits, even filtered graphs, or with `Inverse` or `Undirected` never change the view of nodes. They only change how edges are seen.
So you only ever iterate on all nodes, which I think covers your concerns.

Does this make sense?

Reply:
Oh right, this makes all the sense, unless with have some filtered stuff `nodes` should never be impacted by the absence of use of the `GT` parameter.